crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -82,11 +83,104 @@ pub trait SystemParamFetch<'world, 'state>: SystemParamState {
     unsafe fn get_param(
         state: &'state mut Self,
         system_meta: &SystemMeta,
-        world: &'world World,
+        world: SemiSafeCell<'world, World>,
         change_tick: u32,
     ) -> Self::Item;
 }
 
+/// A non-existent component accessed by systems with params that hold a
+/// [`World`](crate::world::World) reference.
+///
+/// Current list:
+/// - [`Query`](crate::system::Query)
+/// - [`Commands`](crate::system::Commands)
+/// - [`&World`](crate::world::World)
+/// - [`&mut World`](crate::world::World)
+/// - [`&Archetypes`](crate::archetype::Archetypes)
+/// - [`&Bundles`](crate::bundle::Bundles)
+/// - [`&Components`](crate::component::Components)
+/// - [`&Entities`](crate::entity::Entities)
+/// - [`ParamSet`] (if includes any of above)
+#[derive(Component)]
+pub(crate) struct WorldAccess;
+
+/// Marks system as having a param that holds a `&World` reference.
+fn add_shared_world_access(
```

Comment:
I want to record that I really dislike that we use an archetype to record this access.
It makes manual `SystemParam` implementations much more fraught, as demonstrated by the prior `Query<()>` unsoundness (especially since this function is private).

I'd very much like to avoid this, e.g. by making an `get_exclusive` method on `SystemParamState` which gets called if the system is exclusive, and used `Option::take` to get the world reference if required (probably with a default impl which uses `Option::take` and just calls `get`).

Reply:
I changed the way `SemiSafeCell` works based on your feedback. Thanks for all the suggestions. Systems with conflicting params should now always panic during the first conflicting `get_param`, if `init` didn't already.

I removed all the `WorldAccess` stuff since it was no longer essential, but it may be worth having for clearer error messages.